Add release notes for GPU-driven rendering (#2097)
bevyengine/bevy-website: release-content/0.16/release-notes/gpu_driven_rendering.md

## GPU-Driven Rendering

> Over the years, the trend in real-time rendering has increasingly been to move work from the CPU to the GPU. One of the latest developments in this area has been *GPU-driven rendering*, in which the GPU takes a representation of the scene and essentially works out what to draw on its own. While Bevy 0.15 had some support for GPU-driven rendering for meshlets only, Bevy 0.16 features comprehensive support for this technique, including for skinned meshes. This dramatically reduces the amount of CPU time that the renderer needs for larger scenes. It's automatically enabled on platforms that support it; unless your application hooks into the rendering pipeline, upgrading to Bevy 0.16 will automatically enable GPU-driven rendering for your meshes.

Reviewer: ```suggestion
Over the years, the trend in real-time rendering has increasingly been to move work from the CPU to the GPU. One of the latest developments in this area has been *GPU-driven rendering*, in which the GPU takes a representation of the scene and essentially works out what to draw on its own. While Bevy 0.15 had some support for GPU-driven rendering for virtual geometry only, Bevy 0.16 features comprehensive support for this technique, including for skinned meshes. This dramatically reduces the amount of CPU time that the renderer needs for larger scenes. It's automatically enabled on platforms that support it; unless your application hooks into the rendering pipeline, upgrading to Bevy 0.16 will automatically enable GPU-driven rendering for your meshes.
```

We are trying to move away from using `meshlets` and call it virtual geometry instead.

Reviewer 2: +1, I'm planning on renaming the actual code sometime soon as well.

> Over the years, the trend in real-time rendering has increasingly been to move work from the CPU to the GPU. One of the latest developments in this area has been *GPU-driven rendering*, in which the GPU takes a representation of the scene and essentially works out what to draw on its own. While Bevy 0.15 had some support for GPU-driven rendering for meshlets only, Bevy 0.16 features comprehensive support for this technique, including for skinned meshes. This dramatically reduces the amount of CPU time that the renderer needs for larger scenes. It's automatically enabled on platforms that support it; unless your application hooks into the rendering pipeline, upgrading to Bevy 0.16 will automatically enable GPU-driven rendering for your meshes.

Reviewer 2: ```suggestion
Over the years, the trend in real-time rendering has increasingly been to move work from the CPU to the GPU. One of the latest developments in this area has been *GPU-driven rendering*, in which the GPU takes a representation of the scene and essentially works out what to draw on its own. While Bevy 0.15 had support for GPU-driven rendering for virtual geometry only, Bevy 0.16 features support for this technique for the standard mesh rendering path, including for skinned meshes. This dramatically reduces the amount of CPU time that the renderer needs for larger scenes. It's automatically enabled on platforms that support it; unless your application hooks into the rendering pipeline, upgrading to Bevy 0.16 will automatically enable GPU-driven rendering for your meshes.
```

To explain how GPU-driven rendering operates, it's easiest to first describe how CPU-driven rendering works:

1. The CPU determines the objects that are visible, via frustum culling and perhaps occlusion culling.

2. For each such object:

   a. The CPU sends the object's transform to the GPU, possibly in addition to other data such as joint weights.

   b. The CPU tells the GPU where the mesh data is.

   c. The CPU writes the material data to the GPU.

   d. The CPU tells the GPU where textures and other buffers needed to render the objects are (light data, etc.)

   e. The CPU issues a drawcall.

   f. The GPU renders the object.

In contrast, GPU-driven rendering in Bevy works like this:

> 1. The CPU uploads transform information for all objects at once to the GPU.

Reviewer 3: This is the same as for CPU-driven.

Author: I intentionally gloss over the difference between `MeshInputUniform` and `MeshUniform`, and the difference between the GPU-preprocessing and no-GPU-preprocessing paths, the details of which I thought were too complex and out-of-scope for this blog post.

I reworded it to "The CPU supplies a single buffer containing transform information for all objects to the GPU, so that shaders can process many objects at once.".

2. If new objects have been spawned since the last frame, the CPU fills out tables specifying where the mesh data for the new objects are.

> 3. If materials have been modified since the last frame, the CPU uploads information about those materials to the GPU.

Reviewer 3: This is the same as for CPU-driven.

Reviewer 2: Yeah, I feel like the _real_ GPU-driven part is the culling. Stuff like bindless, multidraw, and caching out specialization are all important, and could be covered in separate section, but are not really GPU-driven techniques (I suppose you could argue bindless is).

Author: My view is that "GPU driven" is a combination of many techniques, none of which *on their own* indicate "GPU driven", but when you put them all together they add up to that technique. GPU driven isn't culling, bindless, multidraw, or cached specialization: it's what happens when you *combine* all of those.

4. The CPU creates lists of objects to be rendered this frame. Each object is simply referenced by an integer ID, so these lists are small. The number of lists depends on the size and complexity of the scene, but there are rarely more than 15 such lists even for large scenes.

5. For each such list:

    a. The CPU issues a *single* drawcall.

    b. The GPU processes all objects in the list, determining which ones are truly visible.

    c. The GPU renders each such visible object.

> For large scenes that may have tens of thousands of objects, GPU-driven rendering frequently results in reduction in CPU rendering overhead of 3× or more. It's also necessary for occlusion culling, because of the GPU transform step (5(b) above).

Reviewer 2: ```suggestion
For large scenes that may have tens of thousands of objects, GPU-driven rendering frequently results in a reduction in CPU rendering overhead of 3× or more. It's also necessary for occlusion culling, because of the GPU transform step (5(b) above).
```

Internally, GPU-driven rendering is less a single technique than a collection of techniques. These include:

* *Multi-draw indirect* (MDI), a GPU API that allows multiple meshes to be drawn in a single drawcall, the details of which the GPU provides by filling out tables in GPU memory. In order to use MDI effectively, Bevy uses a new subsystem, the *mesh allocator*, which manages the details of packing meshes together in GPU memory.

  - *Multi-draw indirect count* (MDIC), an extension to multi-draw indirect that allows the GPU to determine the *number* of meshes to draw with minimal overhead.

> * *Bindless resources*, which allow Bevy to supply the textures (and other resources) for many objects as a group, instead of having to bind textures one-by-one on the CPU. These resources are managed by a new subsystem known as the *material allocator*.

Reviewer: I originally expected this release note to contain a bit more details about the bindless feature that mentions the nuance between what we call bindless and what some purist call bindless. But I'm starting to see it's a bit out of scope and might need a separate section. I guess it's also fair to just defer anyone interested in it to go read the docs for it.

Author: Are you referring to the lack of first-class bindless in wgpu? If so, that's really a wgpu-specific issue and wouldn't be sensible to those who aren't intimately familiar with the wgpu API. No other API that I'm aware of has that problem. If we wanted to mention it, I guess we could put something like "wgpu has plans to improve the performance of bindless" in the "future work" section.

Reviewer: Tbh, I don't really know. All I know is I've seen people mention "real bindless" many times and saying that wgpu doesn't support it yet. I don't fully understand what the difference is or if it's even worth mentioning or if there even is one. 

I'm not just referring to discord I've seen it in other places too. I just don't know much about bindless outside the context of bevy.

Reviewer 2: The problem is that wgpu does not support modifying an existing bind group, and recreating binding arrays with large binding arrays is really expensive.

So every time you want to load or unload a new texture, you need to recreate the entire bind group, which takes several milliseconds. Hence why @pcwalton limited bevy to batches of tiny binding arrays, instead of what every other renderer does which is a single bind group of every single texture in the scene.

Author: We don't use tiny binding arrays anymore; I fixed that in https://github.com/bevyengine/bevy/pull/17898. The default slab size is now 2048.

> * *GPU transform and cull*, which allows Bevy to compute the position and visibility of every object from the camera's perspective on the GPU instead of on the GPU.

Reviewer: ```suggestion
* *GPU transform and cull*, which allows Bevy to compute the position and visibility of every object from the camera's perspective on the GPU instead of on the CPU.
```

> * The *retained render world*, which allows the CPU to avoid processing and uploading data that hasn't changed since the last frame. Bevy 0.16 not only retains entities in the render world from frame to frame but also retains many internal rendering data structures, allowing it to avoid most CPU overhead associated with meshes that didn't change from the previous frame.

Reviewer 3: I know cold specialisation is a kind of retention, but I feel like it is worth stating more explicitly that material render pipelines are specialised only when something changes that needs them to be re-specialised.

Author: I split this out into a new bullet point, which I'm calling "cached pipeline specialization" ("cold specialization" is a bit idiosyncratic of a name for a public blog post, IMO).

> * The *retained render world*, which allows the CPU to avoid processing and uploading data that hasn't changed since the last frame. Bevy 0.16 not only retains entities in the render world from frame to frame but also retains many internal rendering data structures, allowing it to avoid most CPU overhead associated with meshes that didn't change from the previous frame.

Reviewer 2: Same thing about not really being GPU-driven. Maybe it's just worth renaming the whole section to "Rendering Optimizations" or something.

> Not all platforms offer full support for GPU-driven rendering; for example, Metal (macOS and iOS) currently only has partial support, and WebGL 2 (the browser) has no support for GPU-driven rendering at all. Some platforms, such as Android phones with certain Mali chips, theoretically support GPU-driven rendering, but in practice driver issues prevent it from working reliably. In these cases, Bevy 0.16 automatically works around the missing functionality, or falls back to CPU-driven rendering if that isn't feasible. This is possible because the Bevy renderer is a unified codebase that supports both CPU- and GPU-driven rendering techniques. That is, there is just one renderer, not a CPU-driven renderer and a GPU-driven renderer.

Reviewer 3: I feel like GPU-driven is such a valuable and important performance feature, it would be great to have a support table here that covers the different feature parts of GPU-driven, the platform/graphics API/GPU feature/limit requirements, and the support for those across windows/linux/macos/android/i*OS/webgpu/webgl2. I am happy to help with this.

Author: I made a support table. Feel free to verify for accuracy.

In most cases, you don't need to do anything special in order for your application to support GPU-driven rendering. There are two main exceptions:

1. Materials with custom WGSL shaders will continue to use CPU-driven rendering by default. In order for your materials to use GPU-driven rendering, you'll want to use the new `#[bindless]` feature on `AsBindGroup`. See the `AsBindGroup` documentation and the `shader_material_bindless` example for more details. If you're using `ExtendedMaterial`, check out the new `extended_material_bindless` example.

2. Applications and plugins that hook into the renderer at a low level will need to be updated to support GPU-driven rendering. The newly-updated `custom_phase_item` and `specialized_mesh_pipeline` examples may prove useful as a guide to do this.

> Bevy's current GPU-driven rendering isn't the end of the story. There's a sizable amount of potential future work to be done:

Reviewer 3: I'd like us to add a bullet point about leveraging various parts of this to avoid rebinding for CPU-driven too. Say that WebGPU got bindless before MDI, then bindless would be valuable for CPU-driven. The mesh allocator was in 0.15, but could be referenced as something that was already done if interesting to do so.

Author: Done.

Reviewer 3: Thanks!!!

> * Bevy 0.16 only supports GPU-driven rendering for the 3D pipeline, but the techniques are equally applicable to the 2D pipeline. Future versions of Bevy could support 2D GPU-driven rendering as well.

Reviewer: ```suggestion
* Bevy 0.16 only supports GPU-driven rendering for the 3D pipeline, but the techniques are equally applicable to the 2D pipeline. Future versions of Bevy will support 2D GPU-driven rendering as well.
```

Maybe I'm being too optimistic, but I already started working towards that.

Reviewer 3: Not too optimistic. We will make that happen for 0.17

Reviewer 3: Also not just 2D, but should be applicable to sprites, UI, gizmos, wireframes, ....

Author: I rephrased to "should".

> * Bevy 0.16 only supports GPU-driven rendering for the 3D pipeline, but the techniques are equally applicable to the 2D pipeline. Future versions of Bevy could support 2D GPU-driven rendering as well.

Reviewer 2: Also UI.

> * Morph targets currently force objects to CPU-driven rendering. This is something we plan to address in the future.

Reviewer: Only entities with morph targets right? As in, other entities will still use gpu driven?

If that's the case, I'd suggest maybe adding something like this
```suggestion
* Morph targets currently force objects to CPU-driven rendering for those objects only. This is something we plan to address in the future.
```

> * We're watching new API features, such as [Vulkan device generated commands] and [Direct3D 12 work graphs], with interest. These would allow future versions of Bevy to offload even more work to the GPU, such as sorting of transparent objects. While figuring out how to unify these disparate APIs in a single renderer will be a challenge, the future possibilities in this space are exciting.

Reviewer 2: You could already do transparent object sorting on the GPU. Write out a buffer of depths and then run a radix sort or something, and then do indirect draws.

What DGC and work graphs really do is stuff like only bind and run the pipeline/shader for materials actually on screen (skipping materials for entities that were entirely culled).

Author: That only works if all your transparent objects use the same shader, which isn't realistic for most apps. The reason you need DGC for real world transparent sorting is that you don't know when to switch shaders until you've done the sorting.

Reviewer 2: Ah yeah this is true.
